### PR TITLE
[Operator Tool] Add additional checks to `VerifyValidatorState`

### PR DESCRIPTION
**Pull request by khiemngo**
## Motivation

This PR aims to improve the implementation of `VerifyValidatorState` by:

- Addressing nits in PR #10098.
- Adding check for consensus key uniqueness.
- Checking if the validator network key held in secure storage matches that registered on-chain.
- Checking if the fullnode network key held in secure storage matches that registered on-chain.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Modify the existing test `test_verify_validator_state` to cover new checks.
Screen Shot 2021-12-29 at 11 09 25 AM

## Related PRs

PR #10098 

## If targeting a release branch, please fill the below out as well

 * Justification and breaking nature (who does it affect? validators, full nodes, tooling, operators, AOS, etc.)
 * Comprehensive test results that demonstrate the fix working and not breaking existing workflows.
 * Why we must have it for V1 launch.
 * What workarounds and alternative we have if we do not push the PR.

### CODE REVIEW

File: config/management/operational/src/validator_state.rs [hunk 1]
```diff
@@ -2,67 +2,123 @@
...
+#[derive(Debug, Serialize)]
 pub struct VerifyValidatorStateResult {
-    /// Check if a validator is in the latest validator set on-chain.
-    pub in_validator_set: Option<bool>,
     /// Check if the consensus key held in secure storage matches
     /// that registered on-chain for the validator.
     pub consensus_key_match: Option<bool>,
+
+    /// If the consensus_key_match is true, check if the key is unique
+    pub consensus_key_unique: Option<bool>,
```
Comment (JoshLind) [inline config/management/operational/src/validator_state.rs:21]: Why do we need the `consensus_key_match` to be true to check this? It seems orthogonal, no? In my mind, the more checks we can do the better, i.e., checks should only be avoided if they aren't possible/don't make sense any more.

File: config/management/operational/src/validator_state.rs [hunk 2]
```diff
@@ -2,67 +2,123 @@
...
+    /// Check if a validator is in the latest validator set on-chain.
+    pub in_validator_set: Option<bool>,
+
+    /// Check if the validator network key held in secure storage matches
+    /// that registered on-chain.
+    pub validator_network_key_match: Option<bool>,
 }
 
 impl VerifyValidatorStateResult {
     pub fn is_state_consistent(&self) -> bool {
```
Comment (JoshLind) [inline config/management/operational/src/validator_state.rs:36]: nit: maybe rename to `is_valid_state`? I worry that `is_state_consistent` doesn't seem broad enough (e.g., to cover the case where we do uniqueness checks?)

File: testsuite/smoke-test/src/operational_tooling.rs [hunk 3]
```diff
@@ -1048,11 +1048,14 @@ async fn test_validator_set() {
 
 #[tokio::test]
 async fn test_verify_validator_state() {
-    let (_env, op_tool, backend, mut storage) = launch_swarm_with_op_tool_and_backend(4).await;
+    let (_env, op_tool, backend, mut storage) = launch_swarm_with_op_tool_and_backend(1).await;
 
     let result = op_tool.verify_validator_state(&backend).unwrap();
     assert_eq!(result.in_validator_set, Some(true));
     assert_eq!(result.consensus_key_match, Some(true));
+    assert_eq!(result.consensus_key_unique, Some(true));
```
Comment (JoshLind) [inline testsuite/smoke-test/src/operational_tooling.rs:1056]: nit: you could assert that `is_state_consistent`/`is_valid_state` returns true here, right? Instead of all of these?

File: config/management/operational/src/validator_state.rs [hunk 4]
```diff
@@ -2,67 +2,123 @@
...
 
 #[derive(Debug, StructOpt)]
 pub struct VerifyValidatorState {
+    #[structopt(flatten)]
+    config: ConfigPath,
     #[structopt(long, required_unless = "config")]
     json_server: Option<String>,
     #[structopt(flatten)]
-    validator_config: diem_management::validator_config::ValidatorConfig,
+    validator_backend: ValidatorBackend,
```
Comment (JoshLind) [inline config/management/operational/src/validator_state.rs:52]: Nice! Much cleaner.

File: config/management/operational/src/validator_state.rs [hunk 5]
```diff
@@ -2,67 +2,123 @@
...
+            consensus_key_match: None,
+            consensus_key_unique: None,
+            fullnode_network_key_match: None,
+            in_validator_set: None,
+            validator_network_key_match: None,
+        };
 
-        // TODO(khiemngo): consider return early if the validator is not in the set
+        // Verify if the validator is in the set.
+        // If it is not, return result and no need to do other checks.
```
Comment (JoshLind) [inline config/management/operational/src/validator_state.rs:78]: See my comment above, but I think we should try to do as many checks as possible (that make sense). This will help when debugging, e.g., think about a validator who has just joined. They may not be in the set yet, but they should still be able to verify that their keys match on-chain, etc, right? Why only wait until they are in the set before allowing them to verify local vs on-chain state

File: config/management/operational/src/validator_state.rs [hunk 6]
```diff
@@ -2,67 +2,123 @@
...
+        // If the consensus key matches, check if it is unique
+        if result.consensus_key_match == Some(true) {
+            result.consensus_key_unique = Some(!validator_infos.iter().any(|vi| {
+                vi.account_address() != &owner_account && vi.consensus_public_key() == &storage_key
+            }));
+        }
+
+        // Check if the validator network key held in secure storage
+        // matches that registered on-chain.
+        let storage_key = storage.ed25519_public_from_private(VALIDATOR_NETWORK_KEY)?;
```
Comment (JoshLind) [inline config/management/operational/src/validator_state.rs:106]: Another key consistency check we should think about adding is that all keys (consensus, validator network and fullnode network) match locally when compared with the validator info of the validator set (i.e., the snapshot of the validator configs taken when a new epoch was started). We can do this in a future commit 😄

To elaborate, you already have the `validator_infos` here, which hold the snapshot of all validator configs when the epoch started. If we find that we are in the current validator set, we should then verify that the keys held in the validator set infos match ours locally. Does this make sense?

**Review by JoshLind — COMMENTED** [review]
Coming along nicely @khiemngo 🚀

File: config/management/operational/src/validator_state.rs (hunk 1 above)
Comment (khiemngo) [reply config/management/operational/src/validator_state.rs:21]: Updated to check `consensus_key_unique` even if consensus key does not match.

You're right that this `consensus_key_unique` check should be orthogonal to `consensus_key_match`. As you mentioned, It is useful when a new node (who is not in the validator set yet) wants to check if its key is unique.
(At first, I only thought about the case where a node wants to check if it is already in the set, and then if its key is unique among the nodes in the set.)

File: config/management/operational/src/validator_state.rs (hunk 2 above)
Comment (khiemngo) [reply config/management/operational/src/validator_state.rs]: Changed to `is_valid_state`.

File: testsuite/smoke-test/src/operational_tooling.rs (hunk 3 above)
Comment (khiemngo) [reply testsuite/smoke-test/src/operational_tooling.rs]: Changed to `assert!(result.is_valid_state());`

This is much shorter! Thank you!

File: config/management/operational/src/validator_state.rs (hunk 5 above)
Comment (khiemngo) [reply config/management/operational/src/validator_state.rs]: Yeah it all makes sense now! I updated to proceed with other checks even if the node is not in the set.

File: config/management/operational/src/validator_state.rs (hunk 6 above)
Comment (khiemngo) [reply config/management/operational/src/validator_state.rs:100]: Added `TODO` for this check.

I will follow up with you to ask some questions about this. Thanks!

File: config/management/operational/src/validator_state.rs [hunk 7]
```diff
@@ -2,67 +2,119 @@
...
+        result.in_validator_set = Some(
+            validator_infos
+                .iter()
+                .any(|vi| vi.account_address() == &owner_account),
+        );
+
+        // Fetch the current on-chain config for this operator's owner.
+        // Check if the consensus key held in secure storage matches
+        // that registered on-chain.
         let validator_config = client.validator_config(owner_account).and_then(|vc| {
```
Comment (khiemngo) [inline config/management/operational/src/validator_state.rs:87]: One thing I'm not sure/want to ask is: even if this node is not in the validator set, but its account is registered/available on-chain, the call `client.validator_config(owner_account)` should be able to return the config (unless there are some unexpected errors (e.g., network))? @JoshLind

**Comment by khiemngo** [comment]
Thanks for the feedback @JoshLind esp. during this holiday week! (I did not expect to receive feedback until after new year 😄 )

File: config/management/operational/src/validator_state.rs (hunk 7 above)
Comment (JoshLind) [reply config/management/operational/src/validator_state.rs:87]: @khiemngo, yeah, exactly. It should still work 😄

File: testsuite/smoke-test/src/operational_tooling.rs [hunk 8]
```diff
@@ -1061,8 +1060,12 @@ async fn test_verify_validator_state() {
     let result = op_tool.verify_validator_state(&backend).unwrap();
     assert_eq!(result.in_validator_set, Some(true));
     assert_eq!(result.consensus_key_match, Some(false));
+    assert_eq!(result.consensus_key_unique, Some(true));
+    assert_eq!(result.validator_network_key_match, Some(true));
+    assert_eq!(result.fullnode_network_key_match, Some(true));
 
     // TODO(khiemngo): consider adding test where the validator is no longer in set
+    // TODO(khiemngo): consider adding test where consensus key is not unique
```
Comment (JoshLind) [inline testsuite/smoke-test/src/operational_tooling.rs:1068]: One way of doing this would be to create a swarm of several nodes, and then manually change the consensus key of one of them to match another. This would probably requiring manually overwriting the consensus key in storage and then calling the `rotate_consensus_key` function of the operator tool 😄

File: config/management/operational/src/validator_state.rs [hunk 9]
```diff
@@ -2,67 +2,119 @@
...
 
-        // TODO(khiemngo): add checks for validator/fullnode network addresses
-        // TODO(khiemngo): add check for key uniqueness
+        Ok(result)
 
-        Ok(VerifyValidatorStateResult {
-            in_validator_set: Some(in_validator_set),
-            consensus_key_match: Some(consensus_key_match),
-        })
+        // TODO(khiemngo): Check if all keys match locally when compared with the validator infos
```
Comment (JoshLind) [inline config/management/operational/src/validator_state.rs:118]: As this function grows in the future, I think we could refactor out these checks into helper methods, so the function doesn't become too large. But, let's iterate 😄

**Review by JoshLind — APPROVED** [review]
Awesome, thanks @khiemngo! LGTM 🚢
